File: algorithms/T/Throne Inheritance/Throne Inheritance.java
class Tree{
    List<Tree>child;
    String name;
    public Tree(String name,List<Tree>child){
        this.name=name;
        this.child=child;
    }
}
class ThroneInheritance {
    private Set<String>death;
    private Tree tree;
    private Map<String,Tree>addtoTree;
    public ThroneInheritance(String kingName) {
      death=new HashSet<>();  
      tree=new Tree(kingName,new ArrayList());
      addtoTree=new HashMap();
      addtoTree.put(kingName,tree);   
    }
    
    public void birth(String parentName, String childName) {
           Tree tmp =addtoTree.get(parentName);
           Tree childtree=new Tree(childName,new ArrayList());
           tmp.child.add(childtree);
            addtoTree.put( childName,childtree); 
    }
    
    public void death(String name) {
        death.add(name);
    }
    
    public List<String> getInheritanceOrder() {
        List<String>ans=new ArrayList<>();
        preOreder(tree,ans,death);
        return ans;
    }
    
    void preOreder(Tree n,List<String>ans,Set<String>death){
        if(n==null)return;
        if(!death.contains(n.name))ans.add(n.name);
         for(Tree name:n.child){
            preOreder(name,ans,death);
        }
    }
}

File: algorithms/T/Throne Inheritance/Throne Inheritance.cpp
class ThroneInheritance {
public:
    ThroneInheritance(string kingName) {
        curr_king = kingName;
    }
    
    void birth(string parentName, string childName) {
        children[parentName].push_back(childName);
    }
    
    void death(string name) {
        dead.insert(name);
    }
    
    void rec(string parent) {
        if (!dead.count(parent)) inheritance.push_back(parent);
        for (auto child : children[parent])
            rec(child);
    }
    
    vector<string> getInheritanceOrder() {
        inheritance = {};
        rec(curr_king);
        return inheritance;
    }
    
private:
    unordered_map<string, vector<string>> children;
    vector<string> inheritance;
    unordered_set<string> dead;
    string curr_king;
};
